cirq/ops/gate_operation.py

```diff
@@ -59,15 +58,23 @@ def with_gate(self, new_gate: 'cirq.Gate') -> 'cirq.Operation':
         return new_gate.on(*self.qubits)
 
     def __repr__(self):
+        if hasattr(self.gate, '_op_repr_'):
+            result = self.gate._op_repr_(self.qubits)
+            if result is not None and result is not NotImplemented:
+                return result
+        gate_repr = repr(self.gate)
+        qubit_args_repr = ', '.join(repr(q) for q in self.qubits)
+        assert type(self.gate).__call__ == raw_types.Gate.__call__
+
         # Abbreviate when possible.
+        dont_need_on = re.match(r'^[a-zA-Z0-9.()]+$', gate_repr)
```

Comment:
Regex is too big a hammer for this since we're not checking any structural properties of the string, but merely that its characters belong to a given set (and that the string is non-empty). I don't think we really need to check whether the string is non-empty, so we can probably just do this:

```python
dont_need_on = all(c.isalnum() or c in '(.)' for c in gate_repr)
```

And if we do need to know that `gate_repr` is non-empty then you can just say

```python
dont_need_on = gate_repr and all(c.isalnum() or c in '(.)' for c in gate_repr)
```